mio/src/lib.rs

```diff
@@ -1,111 +1,145 @@
 extern crate mio;
 extern crate futures;
+extern crate fnv;
 
+use std::hash::BuildHasherDefault;
 use std::collections::HashMap;
 use std::io::{self, ErrorKind, Read, Write};
 use std::net::SocketAddr;
 use std::panic;
 use std::slice;
 use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::mpsc::{channel, TryRecvError};
 
-use futures::{Future, promise, Complete, PollError};
+use fnv::FnvHasher;
 
-pub type IoFuture<T> = Future<Item=T, Error=io::Error>;
+use futures::{Future, Tokens, promise, Complete, Wake, PollError, PollResult};
 
-pub struct Loop {
-    io: mio::Poll,
-    tx: mio::channel::Sender<Message>,
-    rx: mio::channel::Receiver<Message>,
-    next: usize,
-    done: HashMap<usize, Complete<(), io::Error>>,
+pub struct MioEvent {
+    events: mio::EventSet,
+    source: Source,
+    loop_handle: LoopHandle,
 }
 
-enum Message {
-    Wait(Complete<(), io::Error>, mio::EventSet, Arc<mio::Evented + Send + Sync>),
-    Register(Arc<mio::Evented + Send + Sync>),
+impl MioEvent {
+    fn into_future<C: PollCompletion>(self, completion: C) -> MioFuture<C> {
```

Comment:
This seems somewhat similar to:

```rust
impl MioEvent {
    fn as_future<F, T, E>(self, f: F) -> MioFuture<F> 
        where F: FnMut(&Tokens) -> Option<PollResult<T, E>>
    {
        // ...
    }
}
```

That is, I'm not sure if the `PollCompletion` trait is buying us much over `FnMut`?